keyboards/custommk/evo70/keymaps/via/keymap.c

```diff
@@ -0,0 +1,98 @@
+	    KC_F9,   KC_F10,           KC_TRNS, RGB_TOG, RGB_MOD, RGB_RMOD, RGB_VAI, RGB_VAD,  RGB_HUI, RGB_HUD, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,  
+		KC_F11,  KC_F12,  KC_TRNS, KC_TRNS,          KC_TRNS,                    KC_TRNS,  KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS),
+
+	[2] = LAYOUT(
+						  KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,  KC_TRNS, KC_TRNS,  KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, 
+						  KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,  KC_TRNS, KC_TRNS,  KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, 
+		KC_TRNS, KC_TRNS, KC_TRNS,          KC_TRNS, KC_TRNS, KC_TRNS,  KC_TRNS, KC_TRNS,  KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, 
+	    KC_TRNS, KC_TRNS, 		   KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,  KC_TRNS, KC_TRNS,  KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, 
+		KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,          KC_TRNS,                    KC_TRNS,  KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS),
+
+	[3] = LAYOUT(
+						  KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,  KC_TRNS, KC_TRNS,  KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, 
+						  KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,  KC_TRNS, KC_TRNS,  KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, 
+		KC_TRNS, KC_TRNS, KC_TRNS,          KC_TRNS, KC_TRNS, KC_TRNS,  KC_TRNS, KC_TRNS,  KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, 
+	    KC_TRNS, KC_TRNS, 		   KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,  KC_TRNS, KC_TRNS,  KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, 
+		KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,          KC_TRNS,                    KC_TRNS,  KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS),
+
+};
+
+static bool custom_encoder_mode = false;
+
+void set_custom_encoder_mode_user(bool custom_mode) {
+	custom_encoder_mode = custom_mode;
+}
+
+
+keyevent_t encoder_ccw = {
+    .key = (keypos_t){.row = 4, .col = 7},
+    .pressed = false
+};
+
+keyevent_t encoder_cw = {
+    .key = (keypos_t){.row = 4, .col = 10},
+    .pressed = false
+};
+
+
+bool encoder_update_user(uint8_t index, bool clockwise) {
+	if (custom_encoder_mode) {
+	    if (clockwise) {
+	        encoder_cw.pressed = true;
+		    encoder_cw.time = (timer_read() | 1);
+		    action_exec(encoder_cw);
+	    }
+	    else {
+		    encoder_ccw.pressed = true;
+		    encoder_ccw.time = (timer_read() | 1);
+		    action_exec(encoder_ccw);
+	    }
+		return false;
+	}
+	return true;
+}
+
+void matrix_scan_user(void) {
+	if (IS_PRESSED(encoder_ccw)) {
+		encoder_ccw.pressed = false;
+		encoder_ccw.time = (timer_read() | 1);
+		action_exec(encoder_ccw);
+	}
+
+	if (IS_PRESSED(encoder_cw)) {
+		encoder_cw.pressed = false;
+		encoder_cw.time = (timer_read() | 1);
+		action_exec(encoder_cw);
+	}
+}
```

Comment:
I assume this can all be converted to the new encoder map feature now.

Reply:
Looking over the implementation for ENCODER_MAP_ENABLE, in theory the code could be converted to use that feature, but in practice there is some difficulty. ENCODER_MAP_ENABLE enforces the encoder's virtual row to be 253 or 252 (for cw and ccw, respectively), and the column to be 0, 1, etc. (in order) for each encoder. The problem is that we've got over 600 EVO70s already shipped to customers which have been programmed with different matrix assignments--that is, we made use of otherwise unused locations within the normal matrix row/column limits. So for ENCODER_MAP_ENABLE the one encoder would basically occupy key locations:
ccw: row == 252, col == 0
cw:  row == 253, col == 0

But the as-shipped hardware is using:
ccw: row == 4, col == 7
cw:  row == 4, col == 10

The ```keymap_key_to_keycode``` function in qmk_firmware/quantum/dynamic_keymap.c tries to call ```dynamic_keymap_get_encoder``` only if the key position exists outside of the normal row/col range, but unfortunately the shipped hardware positions exist within the normal row/col ranges.

The ```keymap_key_to_keycode``` function in qmk_firmware/quantum/dynamic_keymap.c is strong, so I can't override it with a keyboard-specific implementation that would checks for specific rows and columns other than the ones defined through ENCODER_MAP_ENABLE.

So, while it seems that I can certainly make use of ENCODER_MAP_ENABLE to simplify the implementation for future versions of EVO70, I don't see a good way to implement it such that it will continue to match the behavior of the already-shipped hardware. If I can continue using the current implementation based on callbacks, that would be preferred (otherwise, I would have to require customers to flash updated firmware in the future if they want to use VIA or other dynamic keymap software). Also, if I missed something within the code that would let me use ENCODER_MAP_ENABLE with an arbitrarily assigned row/column for ccw/cw, please let me know.